datumaro/components/project.py

```diff
@@ -2,486 +2,2399 @@
-        return self.select(lambda item: not item.path)
+        return pipeline
 
-    def __iter__(self):
-        for subset in self._subsets.values():
-            yield from subset
+    def make_pipeline(self, target) -> Pipeline:
+        if not target in self:
+            raise UnknownTargetError(target)
 
-    def get_subset(self, name):
-        return self._subsets[name]
+        # a subgraph with all the target dependencies
+        if '.' not in target:
+            target = self.make_target_name(target, self[target].head.name)
 
-    def subsets(self):
-        return self._subsets
+        return self._make_full_pipeline().get_slice(target)
 
-    def categories(self):
-        return self._categories
+class GitWrapper:
+    @staticmethod
+    def import_module():
+        import git
+        return git
 
-    def __len__(self):
-        return sum(len(s) for s in self._subsets.values())
-
-    def get(self, id, subset=None, path=None): # pylint: disable=arguments-differ
-        if path:
-            source = path[0]
-            return self._sources[source].get(id=id, subset=subset)
-        return self._subsets.get(subset, {}).get(id)
-
-    def put(self, item, id=None, subset=None, path=None):
-        if path is None:
-            path = item.path
-
-        if path:
-            source = path[0]
-            # TODO: reverse remapping
-            self._sources[source].put(item, id=id, subset=subset, path=path[1:])
-
-        if id is None:
-            id = item.id
-        if subset is None:
-            subset = item.subset
-
-        item = item.wrap(path=path)
-        if subset not in self._subsets:
-            self._subsets[subset] = self.Subset(self, subset)
-        self._subsets[subset].items[id] = item
-        self._length = None
-
-        return item
-
-    def save(self, save_dir=None, merge=False, recursive=True,
-            save_images=False):
-        if save_dir is None:
-            assert self.config.project_dir
-            save_dir = self.config.project_dir
-            project = self._project
+    try:
+        module = import_module.__func__()
+    except ImportError:
+        module = None
+
+    def _git_dir(self):
+        return osp.join(self._project_dir, '.git')
+
+    def __init__(self, project_dir, repo=None):
+        self._project_dir = project_dir
+        self.repo = repo
+
+        if repo is None and \
+                osp.isdir(project_dir) and osp.isdir(self._git_dir()):
+            self.repo = self.module.Repo(project_dir)
+
+    @property
+    def initialized(self):
+        return self.repo is not None
+
+    def init(self):
+        if self.initialized:
+            return
+
+        repo = self.module.Repo.init(path=self._project_dir)
+        repo.config_writer() \
+            .set_value("user", "name", "User") \
+            .set_value("user", "email", "<>") \
+            .release()
+        # gitpython does not support init, use git directly
+        repo.git.init()
```

Comment:
GitPython has troubles working without an upstream repo. `Repo.init` creates a repo without the `HEAD` and branches and is supposed either for creating a "bare" repo (without a working dir) or for binding to an upstream repo. Probably I just wanted to have exactly the `git init` behaviour without bothering about remotes. Maybe the problem deserves an upstream issue.

```python
r = git.Repo.init('test_repo')
r.index.add('x.txt')
r.commit()
# ValueError: Reference at 'refs/heads/master' does not exist
br = r.create_head('master')
# gitdb.exc.BadName: Ref 'HEAD' did not resolve to an object
```

Reply:
`Repo.init` creates a bit incomplete repo without branches and head, so that some gitpython operations can fail until a first commit is done. So I probably wanted to have exactly `git init` behaviour.